### PR TITLE
[Refactor][Ops] Infer static metadata from forward inputs

### LINKED ISSUE #1679: [API][Phase 1] Input-inferred cleanup for non-Conv/Pool/Attention/MoE ops
Parent: #1674

## Scope

Phase-1 child issue for the input-inferred public Op API cleanup. This phase intentionally excludes:

- Conv / AvgPool, already tracked by #1673
- Attention family
- MoE leaf and broader MoE family

Everything else from the #1674 inventory should be handled here, either directly or by splitting smaller child issues before implementation if a family turns out to need its own design pass.

### Norm family

Candidates:

- `FusedAddLayerNormFwdOp`
- `FusedAddRMSNormFwdOp`
- `AdaLayerNormFwdOp`
- `AdaLayerNormZeroFwdOp`
- `BatchNormFwdOp`, `BatchNormBwdOp`
- `GroupNormFwdOp`, `GroupNormFwdOpNoAffine`
- `InstanceNormFwdOp`, `InstanceNormFwdOpNoAffine`

Infer from tensor metadata where possible:

- `M`, `N`, `C`, batch, and spatial extents from input / affine / running-stat shapes.

Keep explicit:

- `eps`, `training`, `momentum`, `num_groups`, `use_input_stats`, and variant flags.

### Reduction cumulative ops

Candidates:

- `CumsumFwdOp`
- `CumprodFwdOp`

Infer:

- `N` from `x.shape[dim]`
- `M` from remaining dimensions

Keep explicit:

- `dim` and semantic accumulation options.

### Linear Attention / SSM

Candidates:

- `GLAFwdOp`, `GLABwdOp`, `GLADecodeOp`
- `DeltaNetFwdOp`, `DeltaNetBwdOp`, `DeltaNetOp`, `DeltaNetDecodeOp`
- `GatedDeltaNetFwdOp`, `GatedDeltaNetBwdOp`, `GatedDeltaNetOp`, `GatedDeltaNetDecodeOp`
- `Mamba2FwdOp`
- `SSDChunkScanFwdOp`, `SSDChunkStateFwdOp`, `SSDDecodeOp`, `SSDStatePassingFwdOp`
- `DaCumsumFwdOp`

Infer:

- `batch`, `seq_len` / `seqlen`, `heads` / `n_heads`, `dim_k`, `dim_v`, `d_head`, `d_state`, `n_groups`, `num_chunks`, and `chunk_len` where these are available from input/state tensor shapes.

Keep explicit:

- `chunk_size`, state-presence flags, return-state flags, and scheduling/state contracts.

### Elementwise / Dropout

Candidates:

- `DropoutOp`
- `UnaryOp` and generated unary activation ops using `N_total`
- `ReciprocalFwdOp`, `NanToNumFwdOp`
- `FusedGatedOp`
- tensor-shape elementwise variants such as `where`, `clamp`, `masked_fill`, and arithmetic variants that currently encode input-like shapes at construction time

Infer:

- `N_total` from `input.numel()`
- `M`, `N` from `x.shape` for fused gated layouts
- broadcasted output shapes from forward tensor metadata for PyTorch-aligned elementwise APIs

Keep explicit:

- scalar values, `inplace`, `training`, approximation modes, and other semantic flags.

### RoPE family

Candidates:

- `RopeNeoxOp`, `RopeNonNeoxOp`, `RopeLlama31Op`, `RopeYarnOp`, `RopeLongRopeOp`
- `RopeNeoxPositionIdsOp`

Infer:

- `batch`, `seq_len`, `num_heads`, `head_dim`, and `num_tokens` from `x.shape` and `position_ids.shape`.

Keep explicit:

- `layout`, `max_position`, `rotary_dim`, `base`, and scaling/original-position parameters.

### Grouped GEMM / FFT

Candidates:

- `GroupedGemmOp`
- `FFTC2COp`

Infer:

- common `n`, `k`, and FFT length `n` from input tensor metadata.

Treat carefully:

- grouped GEMM `batch_sum` / `batch_count` may be scheduling metadata depending on how `batch_sizes` and offsets are represented.

### Engram / MHC / FP8 indexing utilities

Candidates:

- `EngramGateConvFwdOp`, `EngramGateConvBwdOp`, `EngramDecodeOp`
- `MHCPreOp`, `MHCPostOp`
- `FP8LightingIndexerOp`, `FP8QuantOp`
- `TopkSelectorOp`

Infer:

- B/S/H/D-style dimensions from activation, index, or weight tensor metadata where possible.

Keep explicit:

- algorithmic knobs such as `topk`, convolution/dilation/window parameters, range/cleaning flags, and scheduling metadata not available from tensor metadata.

## Required implementation pattern

For each op/family touched in this phase:

- Prefer constructors that expose only static semantic/config params plus non-derivable specialization params.
- Infer tensor-derived dimensions in `forward()` from tensor metadata.
- Build kernels lazily and cache by the resolved specialization signature.
- Do not read CUDA tensor values to infer shape or scheduling metadata.
- Update manifest `signature.params`, `static_dims`, `shape_rules`, workloads, roofline bindings, and validator expectations.
- Update tests, benchmarks, docs, and in-repo call sites to use the preferred API unless they intentionally test compatibility.
- If compatibility constructors are retained, cover them explicitly and keep docstrings/examples teaching the preferred API.

## Suggested internal sequencing

1. Norm + cumulative ops.
2. Elementwise / Dropout / RoPE utilities.
3. Linear Attention / SSM.
4. Grouped GEMM / FFT / Engram / MHC / FP8 utilities.

## Acceptance criteria

- All included families are either completed in this phase or split into smaller linked issues with clear reasons.
- Preferred APIs no longer require dimensions directly derivable from forward tensor metadata.
- Kernel construction remains lazy and cached by the dimensions that actually specialize the kernel.
- Manifest, validators, tests, benchmarks, and docs agree with the preferred public contract.
- Conv/Pool, Attention, and MoE are not changed as part of this phase except for references or coordination notes.

### PR DESCRIPTION
**Pull request by superAngGao**
## Problem

Operator constructors duplicate shape and dtype metadata already available from `forward()` inputs. Closes #1679.

## Solution

- Infer static tensor metadata from inputs and lazily cache kernels by resolved signature.
- Keep semantic configuration such as `chunk_size`, layout, scale, RoPE settings, and `topk` in constructors.
- Update manifests, tests, and benchmarks for the input-inferred APIs.
- Add validation and hot-path signature reuse for affected regularization, FFT, GEMM, MHC, indexing, RoPE, Mamba/SSD, GLA, and DeltaNet ops.

MoE constructors remain out of scope.

## Validation

- `py_compile`, Ruff, and `git diff --check`
- Manifest schema, shape, dtype, and benchmark validation
- `pytest -q tests/ops -m smoke`: 2342 passed, 5 skipped
- GitHub CI, including GPU smoke: passed

### CODE REVIEW

File: tileops/ops/deltanet.py [hunk 1]
```diff
@@ -279,6 +355,56 @@ def default_kernel_map(self) -> Dict[str, Kernel]:
...
+            raise ValueError("k must match q shape")
+        if v.ndim != 4 or v.shape[:3] != (batch, heads, seq_len):
+            raise ValueError("v must have shape [batch, heads, seq_len, dim_v]")
+        if beta.shape != (batch, heads, seq_len):
+            raise ValueError("beta must have shape [batch, heads, seq_len]")
+        dtype = q.dtype
+        if seq_len % self.chunk_size != 0:
+            raise ValueError(
+                f"seq_len ({seq_len}) must be divisible by chunk_size ({self.chunk_size})"
+            )
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tileops/ops/deltanet.py:381]: Done in ce29287c.

File: tileops/ops/mamba2_fwd.py [hunk 2]
```diff
@@ -206,29 +173,72 @@ def forward(
...
+            raise ValueError("C must have shape [batch, seqlen, n_groups, d_state]")
+        if dt.shape != (batch, seqlen, n_heads):
+            raise ValueError("dt must have shape [batch, seqlen, n_heads]")
+        if A.shape != (n_heads,):
+            raise ValueError("A must have shape [n_heads]")
+        if dt_bias is not None and dt_bias.shape != (n_heads,):
+            raise ValueError("dt_bias must have shape [n_heads]")
+        if initial_states is not None and initial_states.shape != (batch, n_heads, d_head, d_state):
+            raise ValueError("initial_states must have shape [batch, n_heads, d_head, d_state]")
+        dev = x.device
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tileops/ops/mamba2_fwd.py:204]: Done in ce29287c.

File: tileops/ops/mhc.py [hunk 3]
```diff
@@ -51,27 +90,54 @@ class MHCPostOp(Op):
...
+            raise ValueError("MHCPostOp expects x_layer_out/h_post/x_res to be 2D tensors")
+        batch, c_x = x_layer_out.shape
+        if h_post.shape[0] != batch or x_res.shape[0] != batch:
+            raise ValueError("MHCPostOp inputs must have matching batch dimensions")
+        n_expand = h_post.shape[1]
+        if x_res.shape[1] != n_expand * c_x:
+            raise ValueError(
+                f"x_res.shape[1] must equal n_expand * c_x={n_expand * c_x}, got {x_res.shape[1]}"
+            )
+        self.batch = batch
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tileops/ops/mhc.py:142]: Done in ce29287c.

File: tileops/ops/rope.py [hunk 4]
```diff
@@ -595,11 +631,14 @@ def _validate_and_prepare(
...
-        expected = (self.num_tokens, self.num_heads, self.head_dim)
-        if tuple(x.shape) != expected:
-            raise ValueError(f"Expected input shape {expected}, got {tuple(x.shape)}")
+        if x.ndim != 3:
+            raise ValueError("RopeNeoxPositionIdsOp expects input shape [tokens, heads, head_dim]")
+        self.num_tokens, self.num_heads, self.head_dim = x.shape
+        rotary_dim = self.head_dim if self.rotary_dim is None else self.rotary_dim
+        if rotary_dim > self.head_dim:
+            raise ValueError("rotary_dim must not exceed head_dim")
+        self.rotary_dim = rotary_dim
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tileops/ops/rope.py]: Done in ce29287c.

File: tileops/ops/fp8_lightning_indexer.py [hunk 5]
```diff
@@ -13,32 +13,104 @@
...
+        if k_batch != batch or k_dim != index_dim:
+            raise ValueError("index_q and index_k must agree on batch and index_dim")
+        if heads % kv_group != 0:
+            raise ValueError("heads must be divisible by kv_group")
+        if weights.shape != (seq_len, heads):
+            raise ValueError("weights must have shape [seq_len, heads]")
+        if cu_seqlen_ks.shape != (seq_len,) or cu_seqlen_ke.shape != (seq_len,):
+            raise ValueError("cu_seqlen_ks/cu_seqlen_ke must have shape [seq_len]")
+        if index_k_scale is not None and index_k_scale.shape != (batch, seq_len_kv, kv_group):
+            raise ValueError("index_k_scale must have shape [batch, seq_len_kv, kv_group]")
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tileops/ops/fp8_lightning_indexer.py]: Done in ce29287c.

File: tileops/ops/topk_selector.py [hunk 6]
```diff
@@ -13,38 +13,70 @@
...
+
     def forward(self, index_score, starts, ends) -> torch.Tensor:
+        if not index_score.is_cuda:
+            raise ValueError("TopkSelectorOp expects CUDA inputs")
+        if index_score.ndim != 4:
+            raise ValueError("TopkSelectorOp expects index_score shape [B, S, S_kv, G]")
+        if starts.ndim != 2 or ends.ndim != 2:
+            raise ValueError("TopkSelectorOp expects starts/ends shape [B, S]")
+        if starts.dtype != torch.int32 or ends.dtype != torch.int32:
+            raise ValueError("TopkSelectorOp expects int32 starts/ends tensors")
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tileops/ops/topk_selector.py:68]: Done in ce29287c.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
The code changes are clean, but the approval-gate prose cleanup must be completed before merge.

### Cross-cutting concerns
- PR body: remove the `## Review follow-up` section and fold only durable final-state facts into `## Solution` or `## Validation`; the approval gate prohibits per-round fix history in the final PR description.
- Review replies: edit the six author replies on the resolved Gemini threads to outcome-only one-liners such as `Done in ce29287c.`; the current rationale and verification narration is prohibited by the reply-discipline gate.

**Review by Ibuki-wind — APPROVED** [review]
Clean — no issues.
